Sources/TuistKit/Generator/ProjectGroups.swift

```diff
@@ -39,6 +64,33 @@ class ProjectGroups {
         }
     }
 
+    func projectGroup(named name: String) throws -> PBXGroup {
+        guard let group = projectGroups[name] else {
+            throw ProjectGroupsError.missingGroup(name)
+        }
+        return group
+    }
+
+    /// Sorts groups in the following order
+    ///
+    /// - Any additional groups or files added to the main group directly
+    /// - Project specified groups
+    /// - Target specified groups (in the order the targets are defined)
+    /// - Remaining predefined groups (e.g. Frameworks, Products etc...)
+    ///
+    func sort() {
```

Comment:
Not a big deal but what about sorting the groups before adding them to the project? I find it odd that we generate the groups and then expect someone from outside to call the `sort()` method.

Reply:
Agreed this was one thing I didn't like :( - hence why I needed to add one more test to make sure it was covered and not forgotten.

The reasoning here for this is if for example a project declares it would like it files at the root (i.e. filesGroup:`nil` - such that there isn't a `Project` group): 

- The groups are generated first within `ProjectGroups.generate(...)`

```
- Manifest
- Frameworks
- Products
```

- Afterwards the sources are processed in `fileElements.generateProjectFiles(...)` and they append to the `main` group directly 

```
- Manifest
- Frameworks
- Products
- Sources
- Tests
```

Hence why a post `sort()` step is needed. Perhaps I'm overcomplicating it

Reply:
Have you considered creating all the groups from `ProjectGroups.generate` including the ones that are specific to the files generation?

Reply:
creating all the groups in one `generate` sweep is the better solution - this requires a bit of extra setup to create another representation of files / groups and then generate groups from that. A simpler alternative is to use orphaned `PBXGroup` placeholders to host elements and then have a `generateMainGroup()` step - but that has a few gotchas where some helpers on `PBXGroup` rely on being part of a `PBXProj` to function (e.g. `addGroup()` and `addFile()`)

The first option could have some additional benefits: it could also help 
https://github.com/tuist/tuist/issues/258 to support globs

Possibly something along the lines of:

```swift
func structure(from paths: [AbsolutePath]) -> Structure
```

This can then be used 

```swift
func workspaceElements(from structure: Structure) -> [XCWorkspaceDataElement]
func fileElements(from structure: Structure) -> [PBXFileElement] 
```

This requires a bit more thought - so will reduce the scope of the PR and omit the ability to specify `nil` and re-introduce it separately in a different PR.

Additional feedback welcome :)

Reply:
> but that has a few gotchas where some helpers on PBXGroup rely on being part of a PBXProj to function (e.g. addGroup() and addFile())

If I'm not wrong, I don't think we are using those helpers so it should be safe adding the files to those orphaned groups.

> Possibly something along the lines of:

What would the `Structure` type contain? I'm struggling to see how that'd fit into the current generation logic.

Reply:
Will prototype this for us to discuss in a separate issue